### PR TITLE
[train] include grafana panels in train export data

### PR DESCRIPTION
**Pull request by matthewdeng**
# Changes

1. Add new `DashboardPanel` schema to `ExportTrainRunEventData` proto.
    1. Add new fields for `train_run_panels` and `train_worker_panels`.
1. Reorganize `train_dashboard_panels.py` so that panels can be individually accessed.
2. Add ordered list of panels directly when converting `train_run_to_proto`.
3. Update `TRAIN_SCHEMA_VERSION`.
4. Added filters for `ray_train_run_id` and `ray_train_worker_world_rank`.
     1. Did **NOT** add a filter for `ray_train_run_attempt_id` as this information is currently not propagated to the workers. This will be done in a subsequent PR.

# Testing

Ran a simple example
```python
from ray.train.v2.api.data_parallel_trainer import DataParallelTrainer

def train_func():
    pass

trainer = DataParallelTrainer(train_func)
trainer.fit()
```

```
RAY_TRAIN_ENABLE_STATE_TRACKING=1 RAY_TRAIN_V2_ENABLED=1 RAY_enable_export_api_write=1 python repro.py
```

Verified that the panels are included in the export events.
```bash
cat /tmp/ray/session_latest/logs/export_events/event_EXPORT_TRAIN_STATE.log | jq -c 'select(.source_type == "EXPORT_TRAIN_RUN") | {status: .event_data.status, train_run_panels: .event_data.train_run_panels, train_worker_panels: .event_data.train_worker_panels}'
{"status":"INITIALIZING","train_run_panels":[{"id":"3","title":"Train Controller State"},{"id":"1","title":"Checkpoint Report Time"},{"id":"2","title":"Train Controller Operation Time"}],"train_worker_panels":[{"id":"1","title":"Checkpoint Report Time"}]}
{"status":"SCHEDULING","train_run_panels":[{"id":"3","title":"Train Controller State"},{"id":"1","title":"Checkpoint Report Time"},{"id":"2","title":"Train Controller Operation Time"}],"train_worker_panels":[{"id":"1","title":"Checkpoint Report Time"}]}
{"status":"RUNNING","train_run_panels":[{"id":"3","title":"Train Controller State"},{"id":"1","title":"Checkpoint Report Time"},{"id":"2","title":"Train Controller Operation Time"}],"train_worker_panels":[{"id":"1","title":"Checkpoint Report Time"}]}
{"status":"FINISHED","train_run_panels":[{"id":"3","title":"Train Controller State"},{"id":"1","title":"Checkpoint Report Time"},{"id":"2","title":"Train Controller Operation Time"}],"train_worker_panels":[{"id":"1","title":"Checkpoint Report Time"}]}
```

### CODE REVIEW

File: python/ray/train/v2/_internal/state/export.py [hunk 1]
```diff
@@ -86,8 +92,33 @@ def train_run_attempt_to_proto(attempt: TrainRunAttempt) -> ProtoTrainRunAttempt
     return proto_attempt
 
 
+def _to_proto_dashboard_panel(panel: Panel) -> ProtoTrainRun.DashboardPanel:
+    """Convert Dashboard Panel to protobuf format."""
+    proto_panel = ProtoTrainRun.DashboardPanel(
+        id=panel.id,
+        title=panel.title,
+    )
```
Comment (matthewdeng) [inline python/ray/train/v2/_internal/state/export.py:100]: Do we need more information, like description?

File: python/ray/train/v2/_internal/state/export.py [hunk 2]
```diff
@@ -100,6 +131,8 @@ def train_run_to_proto(run: TrainRun) -> ProtoTrainRun:
         start_time_ns=run.start_time_ns,
         end_time_ns=run.end_time_ns,
         controller_log_file_path=run.controller_log_file_path,
+        train_run_panels=train_run_panels_proto,
+        train_worker_panels=train_worker_panels_proto,
```
Comment (matthewdeng) [inline python/ray/train/v2/_internal/state/export.py:135]: Should we emit for all events?

File: python/ray/dashboard/modules/metrics/dashboards/train_dashboard_panels.py [hunk 3]
```diff
@@ -5,55 +5,63 @@
...
 )
 
+CHECKPOINT_REPORT_TIME_PANEL = Panel(
+    id=1,
+    title="Checkpoint Report Time",
+    description="Time taken to report a checkpoint to storage.",
+    unit="seconds",
+    targets=[
+        Target(
+            expr="sum(ray_train_report_total_blocked_time_s{{{global_filters}}}) by (ray_train_run_name, ray_train_worker_world_rank)",
```
Comment (alanwguo) [inline python/ray/dashboard/modules/metrics/dashboards/train_dashboard_panels.py:15]: these don't seem to take in additional filters like worker id or run id?
Unless that's part of the global filters

File: python/ray/train/v2/_internal/state/export.py (hunk 1 above)
Comment (alanwguo) [reply python/ray/train/v2/_internal/state/export.py:100]: You can add description to the panel definitions themselves. not necessary here.

Comment (alanwguo) [reply python/ray/train/v2/_internal/state/export.py:100]: You can add description to the panel definitions themselves. not necessary here.

File: python/ray/train/v2/_internal/state/export.py (hunk 2 above)
Comment (alanwguo) [reply python/ray/train/v2/_internal/state/export.py:135]: let's just do all events for now.

File: src/ray/protobuf/export_api/export_train_state.proto [hunk 4]
```diff
@@ -62,6 +62,19 @@ message ExportTrainRunEventData {
...
+  message DashboardPanel {
+    // Unique identifier for the panel
+    string id = 1;
+    // Display name of the panel
+    string title = 2;
+  }
+
+  // List of TrainRun dashboard panels
+  repeated DashboardPanel train_run_panels = 12;
+  // List of dashboard panels to include for TrainWorkers
```
Comment (alanwguo) [inline src/ray/protobuf/export_api/export_train_state.proto:76]: let's include instructions here in the comments that these panels should be scoped to run_id, worker_id, etc

File: python/ray/train/v2/_internal/state/export.py (hunk 1 above)
Comment (matthewdeng) [reply python/ray/train/v2/_internal/state/export.py:100]: Oh okay what about the title?

File: python/ray/dashboard/modules/metrics/dashboards/train_dashboard_panels.py (hunk 3 above)
Comment (matthewdeng) [reply python/ray/dashboard/modules/metrics/dashboards/train_dashboard_panels.py:15]: Currently we just tag the metrics with the `name` and `world_rank` because those are the most understandable by the user (when viewing/filtering metrics). Is that sufficient to map to for the dashboard, or should we also tag with `run_id`? And for the worker would `world_rank` suffice or do we need `actor_id` as well?

Comment (alanwguo) [reply python/ray/dashboard/modules/metrics/dashboards/train_dashboard_panels.py:15]: no i mean filter by. Should these be filtered to a specific run id or worker id when embedded in the dashboard?

File: python/ray/train/v2/_internal/state/export.py (hunk 1 above)
Comment (alanwguo) [reply python/ray/train/v2/_internal/state/export.py:100]: title is useful still because we show title while the embeds are loading.

File: python/ray/dashboard/modules/metrics/dashboards/train_dashboard_panels.py (hunk 3 above)
Comment (alanwguo) [reply python/ray/dashboard/modules/metrics/dashboards/train_dashboard_panels.py:15]: I think name and world rank will work, if we agree that's what we should pass in.

But name and world rank can be re-used across train runs right? So the metrics embedded on run page wont be scoped only to that specific run anymore. It would be all the runs with that same name

Comment (matthewdeng) [reply python/ray/dashboard/modules/metrics/dashboards/train_dashboard_panels.py:15]: Oh actually youre right, we'd technically want to filter exactly by `actor_id`, since there may be multiple run attempts within a run, and multiple runs with the same run name. What's the right way to make it filterable by `worker_id`/`run_id`/`attempt_id`?

Comment (alanwguo) [reply python/ray/dashboard/modules/metrics/dashboards/train_dashboard_panels.py:15]: do we need to scope to a single run attempt? There's no run attempt page with metrics.

But either way, we need to emit enough labels to be able to identify a specific run, run attempt, or worker.

Then grafana dashboard has to add variables for those labels in the train_grafana_dashboard_base.json. These variables should include an All value with default something like `.+`.
Then, the panels need to filter by those variables like run_id=~"$RunId".

Comment (matthewdeng) [reply python/ray/dashboard/modules/metrics/dashboards/train_dashboard_panels.py:15]: Gotcha, I think scoping it to single run attempt would be most precise, and if you were to want to see all metrics for rank 0 across attempts you'd just go to the train run page. 

Will try this out!

File: python/ray/dashboard/modules/metrics/dashboards/train_dashboard_panels.py [hunk 5]
```diff
@@ -5,55 +5,63 @@
...
 )
 
+CHECKPOINT_REPORT_TIME_PANEL = Panel(
+    id=1,
+    title="Checkpoint Report Time",
+    description="Time taken to report a checkpoint to storage.",
+    unit="seconds",
+    targets=[
+        Target(
+            expr="sum(ray_train_report_total_blocked_time_s{{ray_train_worker_world_rank=~'$TrainWorkerWorldRank', {global_filters}}}) by (ray_train_run_name, ray_train_worker_world_rank)",
```
Comment (alanwguo) [inline python/ray/dashboard/modules/metrics/dashboards/train_dashboard_panels.py:15]: if we add a worker id filter as well, we can scope the graph to a single worker of a single attempt since worker id is randomly generated.

**Review by alanwguo — APPROVED** [review]

File: python/ray/dashboard/modules/metrics/dashboards/train_dashboard_panels.py (hunk 5 above)
Comment (matthewdeng) [reply python/ray/dashboard/modules/metrics/dashboards/train_dashboard_panels.py]: I understand now

<div><br/></div>

File: python/ray/dashboard/modules/metrics/dashboards/train_grafana_dashboard_base.json [hunk 6]
```diff
@@ -97,6 +97,93 @@
...
         "useTags": false
+      },
+      {
+        "allValue": ".+",
+        "current": {
+          "selected": false
+        },
+        "datasource": "${datasource}",
+        "definition": "query_result(count by (ray_train_run_id)(last_over_time(ray_train_worker_group_start_total_time_s{{SessionName=~\"$SessionName\",{global_filters}}}[$__range])))",
+        "description": "Filter queries to specific Ray Train run ids.",
```
Comment (justinvyu) [inline python/ray/dashboard/modules/metrics/dashboards/train_grafana_dashboard_base.json:108]: What is this? Why is this filter only for the `ray_train_worker_group_start_total_time_s`?

File: python/ray/dashboard/modules/metrics/dashboards/train_dashboard_panels.py
Comment (justinvyu) [inline python/ray/dashboard/modules/metrics/dashboards/train_dashboard_panels.py:1]: q: why does this stuff live here and not on the anyscale product side?

File: python/ray/train/v2/_internal/state/export.py [hunk 7]
```diff
@@ -86,8 +92,33 @@ def train_run_attempt_to_proto(attempt: TrainRunAttempt) -> ProtoTrainRunAttempt
...
+        CHECKPOINT_REPORT_TIME_PANEL,
+        CONTROLLER_OPERATION_TIME_PANEL,
+    ]
+    # Ordered list of panels to be displayed in the train worker dashboard
+    train_worker_panels = [CHECKPOINT_REPORT_TIME_PANEL]
+
+    train_run_panels_proto = [_to_proto_dashboard_panel(p) for p in train_run_panels]
+    train_worker_panels_proto = [
+        _to_proto_dashboard_panel(p) for p in train_worker_panels
+    ]
```
Comment (justinvyu) [inline python/ray/train/v2/_internal/state/export.py:120]: all of the panel data (like the actual expr to calculate the metric) gets stripped away in this _to_proto step. Maybe call this `_to_proto_dashboard_panel_metadata` instead?

And make it a `ProtoTrainRun.DashboardPanelMetadata`

Comment (justinvyu) [reply python/ray/train/v2/_internal/state/export.py:120]: Also, why do we need to ship the list of dashboard panel ids as part of the run data? Is this to be able to selectively show a few panels if you don't want to show all of them?

File: python/ray/dashboard/modules/metrics/dashboards/train_grafana_dashboard_base.json (hunk 6 above)
Comment (matthewdeng) [reply python/ray/dashboard/modules/metrics/dashboards/train_grafana_dashboard_base.json:108]: This just determines how to find the values

File: python/ray/dashboard/modules/metrics/dashboards/train_dashboard_panels.py
Comment (matthewdeng) [reply python/ray/dashboard/modules/metrics/dashboards/train_dashboard_panels.py:1]: This is for the Ray Grafana panels

File: python/ray/train/v2/_internal/state/export.py (hunk 7 above)
Comment (matthewdeng) [reply python/ray/train/v2/_internal/state/export.py:120]: This is the contract (for now) on how to say which panels correspond to which view (run, worker). I can rename to panel!

**Review by justinvyu — APPROVED** [review]
